sockeye/training.py

```diff
@@ -281,6 +282,12 @@ def _fit(self,
         """
         metric_train = self._create_eval_metric(metrics)
         metric_val = self._create_eval_metric(metrics)
+        # Track loss for optimizer if needed
+        if isinstance(self.module._curr_module._optimizer, SockeyeOptimizer):
+            if self.config.config_loss.type == C.CROSS_ENTROPY:
+                metric_loss = mx.metric.create("ce")
```

Comment:
is the 'ce' metric normalized ?

Reply:
For CE, the stats are not normalized, but the `get()` calls are.  The update rule is:
```
prob = pred[numpy.arange(label.shape[0]), numpy.int64(label)]
self.sum_metric += (-numpy.log(prob + self.eps)).sum()
self.num_inst += label.shape[0]
```
A call to `get_name_value()` returns `sum_metric / num_inst`.

For SCE with loss normalization, the values summed for `sum_metric` are already divided by batch instances as part of the loss calculation, so we just use `num_inst += 1`.

Reply:
I see, so normalization for 'ce' is with respect to the number of instances (the size of the batch dimension), never w.r.t to the number of valid tokens in the batch (i.e. the number of non-pad tokens in labels). This is consistent to what SoftmaxOutput backpropagates if args.normalize_loss is off (i.e. batch normalization of loss, using rescale_grad=1/batch_size).

If args.normalize_loss is on, SoftmaxOutput from the CrossEntropyLoss is using 'valid' normalization, effectively normalizing by the number of valid label tokens in the batch.

For these Eve metrics, we could use batch normalization as default ( `num_inst += label.shape[0]`) if args.normalize_loss is off, and `num_inst += mx.nd.sum(labels != C.PAD_ID)` if args.normalize_loss is on. 
However, doing that consistent for both CE and SCE metrics would also require overwriting the CE metric. Not sure if we want to do this... In any case, it might be hard to exactly get the same loss from the metric as is used to calculate gradients in the graph.

Reply:
Good point about valid tokens.  CE does call `ravel()` so it normalizes by the total number of tokens in the batch (`shape[0]`), but it doesn't check for padding.  The logic is about 10 lines so it's not terrible to keep our own version.  In fact, it's probably a good idea to keep a Metric version for every Loss we have.

Reply:
Agreed, this is however only necessary for losses that are not returned by the symbol themselve (as is the case for CrossEntropy). For all others, e.g. SmoothedCrossEntropy, which also returns a symbol of the loss value, you can use the `Loss` metric from MXNet: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/metric.py#L976